### PR TITLE
Stop optimizing queries twice

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2368

 # Rationale for this change

Why do something twice when you can do it once.

I see speedup of 8% - 11% in the included criterion benchmark for SQL planning.

# What changes are included in this PR?

- SQL execution no longer optimizes the logical plan before creating the physical plan
- DataFrame execution no longer optimizes the logical plan before creating the physical plan
- Optimization happens once when creating physical plan
- Criterion bench added for SQL planning

# Are there any user-facing changes?

Yes. Users will now see unoptimized plans and maybe we will need to make changes to EXPLAIN before we merge this so that they still have a way to see the optimized plan before execution?

### CODE REVIEW

**Comment by andygrove** [comment]
@alamb @Dandandan @matthewmturner This might be a step backward in UX so I left this as a draft while we discuss.

File: datafusion/core/benches/sql_planner.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            )
+        })
+    });
+
+    c.bench_function("trivial join late columns", |b| {
+        b.iter(|| {
+            plan(
+                ctx.clone(),
+                "SELECT t1.a99, t2.b99  \
+                 FROM t1, t2 WHERE a199 = b199",
```
Comment (alamb) [inline datafusion/core/benches/sql_planner.rs:76]: maybe would be nice to add a `GROUP BY` as well :)

It is great to start getting benchmarks on the planner 👍

File: datafusion/core/src/execution/context.rs [hunk 2]
```diff
@@ -1887,26 +1885,6 @@ mod tests {
...
-        let plan1 = ctx
-            .create_logical_plan("SELECT * FROM (SELECT 1) AS one WHERE TRUE AND TRUE")?;
-
-        let opt_plan1 = ctx.optimize(&plan1)?;
-
-        let plan2 = ctx
-            .sql("SELECT * FROM (SELECT 1) AS one WHERE TRUE AND TRUE")
-            .await?;
-
-        assert_eq!(
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1902]: Maybe we could address the UX by making `to_logical_plan` call optimize?

**Review by alamb — APPROVED** [review]
Thanks @andygrove . 

There were previous tickets / PRs on this topic:
https://github.com/apache/arrow-datafusion/issues/1182
https://github.com/apache/arrow-datafusion/pull/1183
https://github.com/apache/arrow-datafusion/issues/705

As I recall the issue was that we were worried that some of the DataFrame APIs would allow  running unoptimized plans. However, when looking through this API it seems like we always run an optimized plan 🤔  I can't remember what the problem was 

The idea of only optimizing once sounds like a good idea to me

cc @xudong963 and @houqp

File: datafusion/core/src/execution/context.rs (hunk 2 above)
Comment (andygrove) [reply datafusion/core/src/execution/context.rs:1902]: Ah, yes, I like that idea.

**Review by alamb — COMMENTED** [review]
This is great -- thank you @andygrove  

the issue kept coming up so I am glad you finally found a solution!

**Pull request merged**